sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveLateralColumnAlias.scala

```diff
@@ -0,0 +1,222 @@
+            case o: OuterReference
+              if aliasMap.contains(o.nameParts.map(_.head).getOrElse(o.name)) =>
+              val nameParts = o.nameParts.getOrElse(Seq(o.name))
+              val aliases = aliasMap.get(nameParts.head).get
+              aliases.size match {
+                case n if n > 1 =>
+                  throw QueryCompilationErrors.ambiguousLateralColumnAlias(nameParts, n)
+                case n if n == 1 && aliases.head.alias.resolved =>
+                  // Only resolved alias can be the lateral column alias
+                  resolveByLateralAlias(nameParts, aliases.head.alias)
+                    .map(LateralColumnAliasReference(_, nameParts))
+                    .getOrElse(o)
+                case _ => o
+              }
+            case u: UnresolvedAttribute if aliasMap.contains(u.nameParts.head) &&
+              Analyzer.resolveExpressionByPlanChildren(u, p, resolver)
+                .isInstanceOf[UnresolvedAttribute] =>
+              val aliases = aliasMap.get(u.nameParts.head).get
+              aliases.size match {
+                case n if n > 1 =>
+                  throw QueryCompilationErrors.ambiguousLateralColumnAlias(u.name, n)
+                case n if n == 1 && aliases.head.alias.resolved =>
+                  // Only resolved alias can be the lateral column alias
+                  resolveByLateralAlias(u.nameParts, aliases.head.alias)
+                    .map(LateralColumnAliasReference(_, u.nameParts))
+                    .getOrElse(u)
+                case _ => u
+              }
+          }.asInstanceOf[NamedExpression]
+        }
+        val newProjectList = projectList.zipWithIndex.map {
+          case (a: Alias, idx) =>
+            val lcaWrapped = wrapLCAReference(a).asInstanceOf[Alias]
+            // Insert the LCA-resolved alias instead of the unresolved one into map. If it is
+            // resolved, it can be referenced as LCA by later expressions (chaining).
+            // Unresolved Alias is also added to the map to perform ambiguous name check, but only
+            // resolved alias can be LCA
+            aliasMap = insertIntoAliasMap(lcaWrapped, idx, aliasMap)
+            lcaWrapped
+          case (e, _) =>
+            wrapLCAReference(e)
+        }
+        p.copy(projectList = newProjectList)
+    }
+
+    // phase 2: unwrap
+    rewrittenPlan.resolveOperatorsUpWithPruning(
+      _.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE), UnknownRuleId) {
+      case p @ Project(projectList, child) if p.resolved
+        && projectList.exists(_.containsPattern(LATERAL_COLUMN_ALIAS_REFERENCE)) =>
+        // build the map again in case the project list changes and index goes off
+        // TODO one risk: is there any rule that strips off /add the Alias? that the LCA is resolved
+        //  in the beginning, but when it comes to push down, it really can't find the matching one?
```

Comment:
I don't think we will strip top-level alias. But if other rules split `Project`, then an alias may become an attribute in the upper `Project`. Fortunately I'm not aware of such rules.

Reply:
Right, I worried that the alias is turned into an attribute in the Project list (in non-top level, I assume LCA can happen in non-top level?). Grad there isn't such case, but even there is, as this PR does, restore back to `UnresolvedAttribute` should be safe to just throw exception.

Reply:
I can't think of any